base/stats.go

```diff
@@ -0,0 +1,791 @@
+package base
+
+import (
+	"expvar"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	StatViewFormat = "%s.%s"
+
+	NamespaceKey                 = "sgw"
+	ResourceUtilizationSubsystem = "resource_utilization"
+
+	SubsystemCacheKey           = "cache"
+	SubsystemDatabaseKey        = "database"
+	SubsystemDeltaSyncKey       = "delta_sync"
+	SubsystemGSIViews           = "gsi_views"
+	SubsystemReplication        = "replication"
+	SubsystemReplicationPull    = "replication_pull"
+	SubsystemReplicationPush    = "replication_push"
+	SubsystemSecurity           = "security"
+	SubsystemSharedBucketImport = "shared_bucket_import"
+)
+
+type SgwStats struct {
+	GlobalStats     *GlobalStat         `json:"global"`
+	DbStats         map[string]*DbStats `json:"per_db"`
+	ReplicatorStats *ReplicatorStats    `json:"per_replication,omitempty"`
+
+	mutex                     sync.Mutex
+	registeredReplicatorStats sync.Once
+}
+
+func NewSyncGatewayStats() *SgwStats {
+	sgwStats := SgwStats{
+		GlobalStats: &GlobalStat{},
+		DbStats:     map[string]*DbStats{},
+	}
+
+	sgwStats.GlobalStats.initResourceUtilizationStats()
+
+	return &sgwStats
+}
+
+func (s *SgwStats) String() string {
+	bytes, err := JSONMarshalCanonical(s)
+	if err != nil {
+		Errorf("Unable to Marshal SgwStats: %v", err)
+		return "null"
```

Comment:
Is null the correct response in an error scenario (as opposed to empty string)?  Looking at other examples (like expvar.String), it seems like an empty string is returned in an error scenario.

Reply:
If we return an empty string it errors on the expvar page which I don't think is the nicest. For example this is the output in the event of a Marshal error:
![image](https://user-images.githubusercontent.com/3264971/89775222-48bbbc00-daff-11ea-829b-520e5cd6a8ea.png)

If we instead make it return a "null" this is the output (you can't see the full output as my browser rendered it as pretty JSON):

![image](https://user-images.githubusercontent.com/3264971/89775384-96382900-daff-11ea-9830-dc5d1b988c33.png)